primitives/consensus/aura/src/block_proposal.rs

```diff
@@ -0,0 +1,73 @@
+use crate::InherentDigest;
+use futures::FutureExt;
+use sp_consensus::{Environment, Proposer};
+use sp_inherents::InherentData;
+use sp_runtime::traits::Block as BlockT;
+use sp_runtime::{Digest, DigestItem};
+use std::future::Future;
+use std::marker::PhantomData;
+use std::time;
+
+/// Proposer factory for PartnerChainsProposer. Allows passing ID: InherentDigest type parameter.
+pub struct PartnerChainsProposerFactory<B: BlockT, E: Environment<B>, ID> {
```

Comment:
This is a new code. In polkadot-sdk we have a customization in a slot worker: https://github.com/input-output-hk/polkadot-sdk/blob/partnerchains-stable2407/substrate/client/consensus/slots/src/lib.rs#L204-L211

To deal with associated types I had to create both a Proposer factory and a Proposer that does this tiny logic before calling propose.